text/0000-ux-guidelines.md

```diff
@@ -0,0 +1,118 @@
+- Feature Name: ux-guidelines
+- Start Date: 2015-07-30
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This is a conventions RFC to settle and document a number of UX guidelines.
+
+* Error Messages
+* Explain Messages
+* Compiler flags
+
+# Motivation
+
+Good user experience is paramount for happy users. We already have informal
+guidelines, and having them documented makes it easier for new developers to
+start contributing without regressing on user experience.
+
+# Detailed design
+
+This RFC includes a number of unrelated UX guidelines broken down into
+arbitrary subsections. It starts with an overall summary of the goals of
+these guidelines. Ideally, everything after this paragraph should be
+copyable into a `ux-guidelines.md` file into the rustc repo. This RFC is
+specifically not trying to create *new* guidelines - just codify the ones that
+already exist informally.
+
+### Summary
+
+Don't forget the user. Whether human or another program, such as an IDE, a
+good user experience with the compiler goes a long way into making developer
+lives better. We don't want users to be baffled by compiler output or
+learn arcane patterns to compile their program.
+
+### Error, Warning, Help, Note Messages
+
+When the compiler detects a problem, it can emit either an error, warning,
+note, or help message.
+
+An `error` is emitted when the compiler detects a problem that makes it unable
+ to compile the program, either because the program is invalid or the
+ programmer has decided to make a specific `warning` into an error.
+
+A `warning` is emitted when the compiler detects something odd about a
+program. For instance, dead code and unused `Result` values.
+
+A `help` is emitted following either an `error` or `warning` giving extra
+information to the user about how to solve their problem.
+
+A `note` is for identifying additional circumstances and parts of the code
+that lead to a warning or error. For example, the borrow checker will note any
+previous conflicting borrows.
+
+* Write in plain simple English. If your message, when shown on a – possibly
+small – screen (which hasn't been cleaned for a while), cannot be understood
+by a normal programmer, who just came out of bed after a night partying, it's
+too complex.
+* `Errors` and `Warnings` should not suggest how to fix the problem. A `Help`
+message should be emitted instead.
+* `Error`, `Warning`, `Note`, and `Help` messages start with a lowercase
+letter and do not end with punctuation.
+* Error messages should be succinct. Users will see these error messages many
+times, and more verbose descriptions can be viewed with the `--explain` flag.
+That said, don't make it so terse that it's hard to understand.
+* The word "illegal" is illegal. Prefer "invalid" or a more specific word
+instead.
```

Comment:
About a month back I filed an issue about switching `illegal` to `invalid` and another person created a PR within 3 days and within a week, it was merged. It was actually after that that I realized that there's no ux guidelines, so I created a PR for it which eventually turned into this RFC. The rationale is that `invalid` gives just as much meaning as `illegal` and there's no reason to use two words when one word works. Though I can also give a hundred other (not so good) rationales...